packages/framework/esm-react-utils/src/useLocations.ts

```diff
@@ -0,0 +1,236 @@
+    };
+  }, [isLoading, data, isValidating, setSize]);
+
+  return memoizedLocations;
+}
+
+export function usePreviousLoggedInLocations(useLoginLocationTag: boolean, searchQuery: string = '') {
+  const { user } = useSession();
+  const searchParams = new URLSearchParams();
+  searchParams.append('_summary', 'data');
+  const previousLoggedInLocationUuids = useMemo(
+    () => user?.userProperties?.previousLoggedInLocations ?? '',
+    [user?.userProperties?.previousLoggedInLocations],
+  );
+  searchParams.append('_id', previousLoggedInLocationUuids);
+  if (useLoginLocationTag) {
+    searchParams.append('_tag', 'Login Location');
+  }
+  if (typeof searchQuery === 'string' && searchQuery != '') {
+    searchParams.append('name:contains', searchQuery);
+  }
+
+  const url = previousLoggedInLocationUuids ? `${fhirBaseUrl}/Location?${searchParams.toString()}` : null;
+
+  const { data, error, isLoading, mutate } = useOpenmrsSWR<LocationResponse>(url);
+
+  const memoisedResults = useMemo(() => {
+    const indices = {};
+    previousLoggedInLocationUuids?.split(',').forEach((locationUuid, indx) => {
+      indices[locationUuid] = indx;
+    });
+    return {
+      previousLoggedInLocations: data?.data?.entry?.sort((a, b) => indices[a.resource.id] - indices[b.resource.id]),
+      error,
+      isLoadingPreviousLoggedInLocations: isLoading,
+      mutatePreviousLoggedInLocations: mutate,
+    };
+  }, [data, isLoading, error, mutate, previousLoggedInLocationUuids]);
+
+  return memoisedResults;
+}
+
+export function useFetchDefaultLocation(useLoginLocationTag: boolean, searchTerm?: string) {
+  const { user } = useSession();
+  const defaultLocationUuid = user?.userProperties?.defaultLocation ?? '';
+  const [isDefaultLocationValid, setIsDefaultLocationValid] = useState(false);
+  let urlSearchParameters = new URLSearchParams();
+  if (useLoginLocationTag) {
+    urlSearchParameters.append('_tag', 'Login Location');
+  }
+  urlSearchParameters.append('_id', defaultLocationUuid);
+  if (searchTerm) {
+    urlSearchParameters.append('name:contains', searchTerm);
+  }
+  const url = defaultLocationUuid ? `${fhirBaseUrl}/Location?${urlSearchParameters.toString()}` : null;
+  const { data, error, isLoading } = useOpenmrsSWR<LocationResponse>(url);
+
+  useEffect(() => {
+    // We can only validate a locationUuid if there is no search term filtering the result any more.
+    if (!searchTerm) {
+      setIsDefaultLocationValid(data ? data?.ok && data?.data?.total > 0 : false);
+    }
+  }, [data?.data?.total, data?.ok, searchTerm]);
+
+  const results = useMemo(
+    () => ({
+      isDefaultLocationValid,
+      defaultLocationArr: isDefaultLocationValid ? data?.data?.entry : [],
+      error,
+      isLoadingDefaultLocation: isLoading,
+    }),
+    [isDefaultLocationValid, data?.data?.entry, error, isLoading],
+  );
+  return results;
+}
+
+export function useLocations(useLoginLocationTag: boolean, count: number = 0, searchQuery: string = '') {
```

Comment:
Since this only pertains to login locations. I'm also unclear on why we need this in the framework itself and not as part of the login app, which is where I think it belongs (there are other uses to locations; these should not be default restricted to login locations and they should not have the same logic that we're using here to surface locations; in other cases where we need locations, it makes sense to surface the location the user is logged into, but not the _previous_ locations they have logged into).

```suggestion
export function useLoginLocations(useLoginLocationTag: boolean, count: number = 0, searchQuery: string = '') {
```

Reply:
> it makes sense to surface the location the user is logged into, but not the previous locations they have logged into

I actually thought that this case you mentioned above would have to handled in other places, say location dropdown somewhere else in UI, to use both default(logged in) and previous logged in locations to show at the top, hence using the same hook everywhere else rather than creating this again and again.

Should I move this from framework to login app again?

Also, we can pass the `useLoginLocationTag` as true/false, hence I believe `useLocations` is a better name for handling both the cases. Please confirm on this as well.
Thanks!

Reply:
So a couple of things here:

1. Locations have an entire tagging system, which are intended for a sort of "folksonomy" of locations. That is, it might make sense (for other applications) to tag all pharmacy locations as "Pharmacy" and then, say, in the dispensing app, load those "Pharmacy" locations as the list of valid locations a user can dispense from. The upshot of this, I think, is that a framework-wide `useLocations()` hook would take an arbitrary `string` for tags rather than a boolean to indicate whether to use a single tag or not.
2. I'm not convinced that previous logged in locations are the right thing to display for anything other than previous logged-in locations. For example, if I am logged into Clinic 1 which has the sub-locations Room A, Room B, and Room C, it might make more sense for a form (like say the start visit form) to present me first with Room A, Room B, and Room C, since those are the sub-locations of the place I am currently working.

Both of these kind of speak against this being a useful enough function to include in the framework, especially with the name `useLocations()`.

Reply:
What I'm getting at isn't that we shouldn't have a `useLocations()` hook in the framework. It's that that hook should have different properties from the hook that is specific to login locations. It would be possible to implement a `useLoginLocations()` hook in terms of a generic `useLocations()` hook, but implementing a generic `useLocations()` hook cannot be easily done from a hook that implements logic that mostly makes sense in the context of login locations.